src/DataStreams/TTLBlockInputStream.cpp

```diff
@@ -153,112 +153,130 @@ void TTLBlockInputStream::readSuffixImpl()
         LOG_INFO(log, "Removed {} rows with expired TTL from part {}", rows_removed, data_part->name);
 }
 
-void TTLBlockInputStream::removeRowsWithExpiredTableTTL(Block & block)
+static ColumnPtr extractRequieredColumn(const ExpressionActions & expression, const Block & block, const String & result_column)
 {
-    auto rows_ttl = metadata_snapshot->getRowsTTL();
+    if (block.has(result_column))
+        return block.getByName(result_column).column;
 
-    rows_ttl.expression->execute(block);
-    if (rows_ttl.where_expression)
-        rows_ttl.where_expression->execute(block);
+    Block block_copy;
+    for (const auto & column_name : expression.getRequiredColumns())
+        block_copy.insert(block.getByName(column_name));
 
-    const IColumn * ttl_column =
-        block.getByName(rows_ttl.result_column).column.get();
+    expression.execute(block_copy);
+    return block_copy.getByName(result_column).column;
+}
 
-    const IColumn * where_result_column = rows_ttl.where_expression ?
-        block.getByName(rows_ttl.where_result_column).column.get() : nullptr;
+void TTLBlockInputStream::executeRowsTTL(Block & block)
+{
+    auto rows_ttl = metadata_snapshot->getRowsTTL();
+    auto ttl_column = extractRequieredColumn(*rows_ttl.expression, block, rows_ttl.result_column);
```

Comment:
Actually fix is here. Previously there was a conflict between calculated columns of primary key and column of TTL expression. 

For example, if we have `ORDER BY toStartOfHour(d)` and `TTL toStartOfHour(d) + INTERVAL 1 HOUR`, in block there will be column `toStartOfHour(d)`, and column with the same name will appear in the middle of computations of TTL expression. There is no way to easily check, that one action of expression is already executed.

Reply:
Why not add this message to the comment of `extractRequieredColumn`?)